### DISCUSSION TITLE
How to solicit a task selection from the user?

### DISCUSSION THREAD

**Discussion by drok** [microsoft/vscode-discussions#943] [Extension Development QnA] [answered]
Hi,

I would like my extension to obtain from the user a choice of a task. I would like the Task object returned to my extension, rather than run directly.

I noticed there is a `TasksQuickAccessProvider` which seems to be related to what I need, but I would love some guidance as to how to implement access to this quick picker.

I think the solution I'm looking for takes a form like...
```
let selectedTask : vscode.Task = await vscode.tasks.selectTask();
```

Obviously I'd prefer not to reimplement or duplicate functionality that already exists.

My extension implements a taskProvider, through which it publishes a executable tests that the user can run from the pallete, or add to their tasks.json. They can also configure my extension to run a specified task (eg, "build xxx in directory yyy" as a prerequisite for some other task). The tasks created by my extension contain data which can be used to launch them under a debugger, or executed directly. I'd like the user to be able to debug tasks of their choice, so implemented a `debugConfigurationProvider` which, when a debug session starts, would prompt the user to select the task to be debugged.

Thank you!

**Comment by theReynald** [comment]
Hi Drok.

How about trying these steps:

1. **Fetch All Tasks:** First, you need to fetch all available tasks. You can do this using vscode.tasks.fetchTasks() which will return a promise that resolves to an array of Task objects.

2. **Show Quick Pick to User:** Once you have the list of tasks, you can display them to the user using the **vscode.window.showQuickPick** method. This method takes an array of items (or strings) to show in the picker, and returns the selected item.

3. **Filter Tasks Relevant to Your Extension:** Since your extension implements a **TaskProvider**, you might want to filter the list of tasks to only include those provided by your extension, based on some criteria like a specific task type or source.

4. **Handle User Selection:** When the user selects a task, handle this selection accordingly. If the user cancels the selection, the returned value will be **undefined**.

Here's a basic implementation of the above steps:
```javascript

const vscode = require('vscode');

async function selectTask() {
    // Fetch all tasks
    const tasks = await vscode.tasks.fetchTasks();

    // Filter tasks if necessary, for example, tasks provided by your extension
    const myExtensionTasks = tasks.filter(task => task.source === 'YourExtensionSource');

    // Map tasks to QuickPick items
    const items = myExtensionTasks.map(task => ({
        label: task.name,
        task: task
    }));

    // Show Quick Pick and get the selected item
    const selected = await vscode.window.showQuickPick(items, {
        placeHolder: 'Select a task to debug'
    });

    // Return the selected task or undefined
    return selected ? selected.task : undefined;
}

// Usage
async function main() {
    let selectedTask = await selectTask();
    if (selectedTask) {
        // Handle the selected task, e.g., launch it under a debugger
        console.log(`Selected Task: ${selectedTask.name}`);
    } else {
        console.log('No task was selected');
    }
}

main();
``` 

Let us know if that helps.

**Comment by drok** [comment]
Thank you @theReynald for the response.

The emphasis in my request is on preserving existing task selection functionality, and not merely selecting a task. To be more specific, I would like the user to be shown:
  * "recently" used tasks more prominently,
  * grouping by "build tasks" or "test tasks",
  * icons / additional description text
  * cog-icon link to customize the task in tasks.json before returning it.
  * any future capabilities that the built-in task selection UI may implement (I don't want my selection UI to appear outdated compared to the built-in)
  * any capabilities implemented in other extensions that might customize the task of "selecting a task". I don't know if it's possible at this time for other extension to implement a more functional task discovery UI, but when/if that will be an option, I'd like my extension to reap the benefit.

By avoiding re-inventing the wheel, I seek to benefit from future improvements to the "wheel" as it exists today and in the future.

How do I call the built-in task picker `TasksQuickAccessProvider` ?

Would I need to first design a PR for the vscode team to implement such that the functionality is exposed as part of the extension APIs ? Or is the functionality already exposed in a non-obvious way?

Thanks again

**Comment by isidorn** [comment]
Not sure if possible. @alexr00 knows better and might be able to help

**Comment by alexr00** [comment]
I don't think we have a way to do that.  @theReynald's suggestion is the way to do this. @meganrogge FYI.

**Comment by abrilurena** [comment]
Hi, @drok! It's been a while, I just wanted to check in and see if you still need any help with this?

**Reply by drok** [reply, marked as answer]
Hi @abrilurena ... thank you for touching base.

Yes, the need remains. @theReynald kindly showed a basic, generic example based on `showQuickPick`, but this example doesn't speak to the thrust of the problem, which is how to present the user with a UI experience consistent with VSCode's built in task picker. The example would work 'in a pinch', and I'm not a pinch.

I think the main issue is that the `TasksQuickAccessProvider` is not as yet exposed as an API that extensions can call.

I'm postponing the (non-essential) feature that relies on a user-selected task, until it can be a first-class citizen.

**Reply by burkeholland** [reply]
Marking this as the answer for now based on the info in this thread. Please feel free to change this if you feel it is incorrect.
